### PR TITLE
New Extension: Key Press 

### PR DESCRIPTION
**Pull request by TheGemDev**
An extension that checks how many times a keyboard key was pressed in a period of time ( maximum delay ), works just like the double click extension but for keyboard keys.

image

### CODE REVIEW

**Comment by tristanbob** [comment]
Does this also have a function to detect how long a key has been depressed?  (for charging spells/attacks)

Condition: "Button X has been pressed for Y seconds"

I also wonder if it should have an expression that shows the percent complete?

Action: Start an X second long charge named "SuperAttack"

Expression: ChargePercent("SuperAttack")

Condition: "SuperAttack" is charging
Condition: "SuperAttack" is fully charged

Action: Cancel charging "SuperAttack" (the user can use this for key released, or perhaps if the player gets hit while charging)

**Comment by 4ian** [comment]
Just commenting to say that this is a great idea and should simplify the work of many people if this can be flexible enough 👍

**Comment by TheGemDev** [comment]
> Does this also have a function to detect how long a key has been depressed? (for charging spells/attacks)
> 
> Condition: "Button X has been pressed for Y seconds"
> 
> I also wonder if it should have an expression that shows the percent complete?
> 
> Action: Start an X second long charge named "SuperAttack"
> 
> Expression: ChargePercent("SuperAttack")
> 
> Condition: "SuperAttack" is charging Condition: "SuperAttack" is fully charged
> 
> Action: Cancel charging "SuperAttack" (the user can use this for key released, or perhaps if the player gets hit while charging)

Sure, i can try adding a feature like that.

**Comment by TheGemDev** [comment]
this works

image

**Comment by TheGemDev** [comment]
I was wondering what data type is required here?. the parameter type is a "keyboard key", but typing the key as a string doesn't seem to work.

image

**Comment by D8H** [comment]
> I was wondering what data type is required here?. the parameter type is a "keyboard key", but typing the key as a string doesn't seem to work.

The screenshot doesn't show the action sentence. I don't have enough context to answer.

**Comment by TheGemDev** [comment]
Okay, I need to take a look at the extension again, I will send an update later in the day.

**Comment by tristanbob** [comment]
> Okay, I need to take a look at the extension again, I will send an update later in the day.

Do you want to work on this more?  It is probably close to being merged when you are ready.

**Comment by gemmomoh** [comment]
yes, i will still work on it.

**Comment by gemmomoh** [comment]
@D8H This was my question, in order to get the key pressed duration I need to get the key user is checking for I used this 
image
image

the parameter type is a keyboard key 
image

but it still returns an error.

**Comment by D8H** [comment]
> @D8H This was my question, in order to get the key pressed duration I need to get the key user is checking for I used this 
> 
> the parameter type is a keyboard key
> 
> but it still returns an error.

It's really hard to follow with the events being cut.
This parameter type don't accept an expression. If you need call private actions use a text parameter.

**Comment by gemmomoh** [comment]
okay

**Comment by TheGemDev** [comment]
image

**Comment by D8H** [comment]
oh, ok, I guess the key parameter type can't be used for expression parameter. The UI should probably not show it in the list in this case.
